File: backjoon/bfs/_1260.py
#https://www.acmicpc.net/problem/1260


node_num, line_num, start_node = map(int, input().split(" "))

matrix = [[0]*(node_num+1) for _ in range(node_num+1)]


for _ in range(line_num):
    tmp = list(map(int, input().split(" ")))
    matrix[tmp[0]][tmp[1]]=1
    matrix[tmp[1]][tmp[0]]=1
    
def dfs(matrix, start):
    visit = []
    stack = [start]
    
    while stack:
        node = stack.pop()
        if node not in visit:
            visit.append(node)
            tmp=[]
            for i in range(len(matrix)):
                if matrix[node][i]==1:
                    tmp.append(i)
            stack.extend(sorted(tmp, reverse=True))
            
    return visit
    
    
def bfs(matrix, start):
    visit = []
    queue = [start]
    
    while queue:
        node = queue.pop(0)
        if node not in visit:
            visit.append(node)
            tmp=[]
            for i in range(len(matrix)):
                if matrix[node][i]==1:
                    tmp.append(i)
            queue.extend(sorted(tmp))
    return visit
                

print(' '.join(map(str, dfs(matrix,start_node))))
print(' '.join(map(str, bfs(matrix,start_node))))

File: baekjoon/graph/1260.js
// https://www.acmicpc.net/problem/1260

const inputs = require("fs").readFileSync("./input.txt").toString().trim().split('\n');

function solution(inputs) {
  let result = [];
  const [N, M, V] = inputs.shift().split(' ').map(Number);
  const edges = inputs.map((input) => input.split(' ').map(Number));
  const graph = convertEdgesIntoGraph(N, edges).map((nodes) => nodes.sort((a, b) => a - b));

  // dfs
  const stack = [...graph[V].slice().reverse()];
  let visit = [V];
  while (stack.length) {
    const nextNode = stack.pop();
    if (visit.includes(nextNode)) continue;

    visit.push(nextNode);
    stack.push(...graph[nextNode].slice().reverse());
  }

  result.push(visit);

  // bfs
  const queue = [...graph[V].slice()];
  visit = [V];
  while (queue.length) {
    const nextNode = queue.shift();
    if (visit.includes(nextNode)) continue;

    visit.push(nextNode);
    queue.push(...graph[nextNode].slice());
  }

  result.push(visit);

  return result;
}

function convertEdgesIntoGraph(nodeCount, edges, hasDirection = false) {
  const graph = Array.from(Array(nodeCount + 1), () => []);

  for (let [from, to] of edges) {
    graph[from].push(to);

    if (!hasDirection) {
      graph[to].push(from);
    }
  }

  return graph;
}

solution(inputs).map((result) => console.log(result.join(' ')))
